src/Evaluator.hs

```diff
@@ -50,313 +450,217 @@ evalErrorText (EvalErrorCase loc val) =
-          withExtendedEnv nx x (ValueSyntax v) $
-          withExtendedEnv nxs xs (ValueSyntax (Syntax (Stx scs loc (List vs)))) $
-          eval rhs
+          let mkEnv = extend nx x (ValueSyntax v)
+                    . extend nxs xs (ValueSyntax (Syntax (Stx scs loc (List vs))))
+          in step $ Down (unCore rhs) (mkEnv env) k
         _ -> next
-    match next (SyntaxPatternList xs) rhs =
-      \case
+    match next (SyntaxPatternList xs) rhs scrutinee env k =
+      case scrutinee of
         (ValueSyntax (Syntax (Stx _ _ (List vs))))
           | length vs == length xs ->
-            withManyExtendedEnv [(n, x, (ValueSyntax v))
-                                | (n,x) <- xs
-                                | v <- vs] $
-            eval rhs
+            let vals = [ (n, x, ValueSyntax v)
+                       | (n,x) <- xs
+                       | v     <- vs
+                       ]
+            in step $ Down (unCore rhs) (vals `extends` env) k
         _ -> next
-    match _next SyntaxPatternAny rhs =
-      const (eval rhs)
+    match _next SyntaxPatternAny rhs _scrutinee env k =
+      step $ Down (unCore rhs) env k
+
+doDataCase :: SrcLoc -> Value -> [(ConstructorPattern, Core)] -> VEnv -> Kont -> EState
+doDataCase loc v0 [] env kont = Er (EvalErrorCase loc v0) env kont
+doDataCase loc v0 ((pat, rhs) : ps) env kont =
+  match (doDataCase loc v0 ps env kont) (\newEnv -> step $ Down (unCore rhs) newEnv kont) [(unConstructorPattern pat, v0)]
+  where
+    match
+      :: EState {- ^ Failure continuation -}
+      -> (VEnv -> EState) {- ^ Success continuation, to be used in an extended environment -}
+      -> [(ConstructorPatternF ConstructorPattern, Value)] {- ^ Subpatterns and their scrutinees -}
+      -> EState
+    match _fk sk [] = sk env
+    match fk sk ((CtorPattern ctor subPats, tgt) : more) =
+      case tgt of
+        ValueCtor c args
+          | c == ctor ->
+            if length subPats /= length args
+              then error $ "Type checker bug: wrong number of pattern vars for constructor " ++ show c
+              else match fk sk (zip (map unConstructorPattern subPats) args ++ more)
+        _otherValue -> fk
+    match fk sk ((PatternVar n x, tgt) : more) =
+      match fk (sk . extend n x tgt) more
+
+-- -----------------------------------------------------------------------------
+-- Top level API
+
+evaluate :: Core -> Either EState Value
+evaluate =  evaluateIn mempty
+
+evaluateIn :: VEnv -> Core -> Either EState Value
+evaluateIn e = yield . until final step . start e . unCore
+
+evaluateWithExtendedEnv :: VEnv -> [(Ident, Var, Value)] -> Core -> Either EState Value
+evaluateWithExtendedEnv env exts = evaluateIn (inserter exts)
+  where
+    inserter = foldl' (\acc (n,x,v) -> Env.insert x n v acc) env
+
+-- TODO DYG: Move to separate module
+projectError :: EState -> EvalError
+projectError (Er err _env _k) = err
+projectError _                = error "debugger: impossible"
```

Comment:
I see that `repl/Main.hs` has
```haskell
case evaluateIn ... of
  Left evalErr -> print $ erroneousValue $ projectError evalErr
  Right val    -> ...
```
so is this "impossible" in the sense that when `evaluateIn` returns a `Left eState`, that `EState` is always an `Er`? In that case, wouldn't it make more sense to return `Left err` or `Left (err, _env, _k)`, to make impossible states unrepresentable?

Reply:
Yes that would make more sense, but I specifically left it this way because this state will become important for the debugger. The debugger will be typed `debugger :: EState -> DebugM EState` or something like that. But the key point is that the debugger should be able to input `EState` which is actually at `Er ...` and return an `EState` that is not in `Er`. So I left as much information as possible until the very last projection or pretty printing because when `evaluateIn` returns a `Left (bunch of state)` the debugger could take that state and then resume the computation. Perhaps I should add a TODO in a comment

Reply:
Thanks for the explanation, that makes sense. But the call to `error` still seems wrong. I can see several possible solutions:

1. Add a precondition to `projectError`'s documentation, e.g. "should only be called on the `EState` inside a `Left` returned by one of the `evaluateX` functions.
2. Have the `evaluateX` functions return an opaque representation of the error, let's say `ErrState`, containing `(err, env, k)`. Then, provide a total function `projectError :: ErrState -> EvalError` and a total function `ErrState -> EState`, so that the resulting `EState` can be given to `debugger :: EState -> DebugM EState`.
3. Have the `evaluateX` functions return `Left (err, estate)`, so they can both display the error and resume from the `estate`, without knowing that the two are related under the hood.